### PR TITLE
fixed Di abstractclassbug

### LINKED ISSUE #191: DI: Consumer class is not equal to the Producer class
The Consumer should not be the class of the service, it should be abstract the location where a provider is injected, so a Provider can have multiple consumers. This is required because of the bug represented by the [failing test](https://github.com/AmyAssist/Amy/commit/4ddf0c95d68d3aa9b0e2d7bb4e55697ee4ca3f64#diff-7f5fc89b7a5e9e1f7d91e0a6b13992a9R140)

### PR DESCRIPTION
**Pull request by Legion2**
removed ServiceFunction
changed ServiceProvider - ServiceConsumer relation to one to many
fixed #191

Pull request status:
- [x] Tests added/updated
- [x] Doku required and added
- [x] Ready for review

### CODE REVIEW

File: core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/configuration/PropertiesProvider.java [hunk 1]
```diff
@@ -40,25 +41,31 @@
...
+
+	private ServiceConsumer<ConfigurationLoader> consumer;
+
+	public PropertiesProvider() {
+		this.consumer = ConsumerFactory.build(null, Util.serviceDescriptionFor(ConfigurationLoader.class));
+	}
 
 	@Override
-	public Properties getService(Map<ServiceDescription<?>, ServiceFactory<?>> resolvedDependencies,
+	public Properties getService(Map<ServiceConsumer<?>, ServiceFactory<?>> resolvedDependencies,
```
Comment (FlxB2) [inline core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/configuration/PropertiesProvider.java]: java doc comment missing

File: di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java [hunk 2]
```diff
@@ -92,13 +92,13 @@ public void loadServices() {
...
 	@Override
-	public synchronized <T> void register(Class<T> serviceType, ServiceFunction<T> serviceFunction) {
+	public synchronized <T> void register(Class<T> serviceType, ServiceProvider<T> serviceFunction) {
 		this.register(Util.serviceDescriptionFor(serviceType), serviceFunction);
 	}
 
 	@Override
 	public synchronized <T> void register(ServiceDescription<T> serviceDescription,
-			ServiceFunction<T> serviceFunction) {
+			ServiceProvider<T> serviceFunction) {
```
Comment (FlxB2) [inline di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java]: java doc comments missing

File: di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java [hunk 3]
```diff
@@ -241,18 +239,21 @@ private boolean hasServiceOfType(ServiceDescription<?> serviceDescription) {
 
 	}
 
+	private <T> void typeSaveResolve(@Nonnull ServiceConsumer<T> dependency, @Nonnull Deque<ServiceProvider<?>> stack,
```
Comment (FlxB2) [inline di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java]: java doc comments missing, same thing in other methods here

Comment (Legion2) [reply di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java]: this is a private method

File: di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java (hunk 2 above)
Comment (Legion2) [reply di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java]: the javadoc is in the interface

File: core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/configuration/PropertiesProvider.java (hunk 1 above)
Comment (Legion2) [reply core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/configuration/PropertiesProvider.java]: the javadoc is in the interface

File: di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java (hunk 3 above)
Comment (FlxB2) [reply di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java]: private methods have to be documented too

File: core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/configuration/PropertiesProvider.java (hunk 1 above)
Comment (FlxB2) [reply core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/configuration/PropertiesProvider.java]: link it, and comment on the implementation. We need more comments in my opinion.

File: di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java (hunk 3 above)
Comment (Legion2) [reply di/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/di/DependencyInjection.java]: ok

File: core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/configuration/PropertiesProvider.java (hunk 1 above)
Comment (Legion2) [reply core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/configuration/PropertiesProvider.java]: `@Override`

**Pull request merged**